polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/EntityCacheGrantManager.java

```diff
@@ -0,0 +1,217 @@
+      @NotNull PolarisCallContext callCtx,
+      @NotNull PolarisEntityCore grantee,
+      @Nullable List<PolarisEntityCore> catalogPath,
+      @NotNull PolarisEntityCore securable,
+      @NotNull PolarisPrivilege privilege) {
+    try {
+      return delegateGrantManager.grantPrivilegeOnSecurableToRole(
+          callCtx, grantee, catalogPath, securable, privilege);
+    } finally {
+      LOGGER.debug("Invalidating cache for securable {} and grantee {}", securable, grantee);
+      entityCache.removeCacheEntry(securable);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull PrivilegeResult revokePrivilegeOnSecurableFromRole(
+      @NotNull PolarisCallContext callCtx,
+      @NotNull PolarisEntityCore grantee,
+      @Nullable List<PolarisEntityCore> catalogPath,
+      @NotNull PolarisEntityCore securable,
+      @NotNull PolarisPrivilege privilege) {
+    try {
+      return delegateGrantManager.revokePrivilegeOnSecurableFromRole(
+          callCtx, grantee, catalogPath, securable, privilege);
+    } finally {
+      LOGGER.debug("Invalidating cache for securable {} and grantee {}", securable, grantee);
+      entityCache.removeCacheEntry(securable);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull LoadGrantsResult loadGrantsOnSecurable(
+      @NotNull PolarisCallContext callCtx, long securableCatalogId, long securableId) {
+    EntityCacheLookupResult lookupResult =
+        entityCache.getOrLoadEntityById(callCtx, securableCatalogId, securableId);
+    if (lookupResult == null || lookupResult.getCacheEntry() == null) {
+      return new LoadGrantsResult(BaseResult.ReturnStatus.GRANT_NOT_FOUND, null);
+    }
+    List<PolarisGrantRecord> grantRecords =
+        lookupResult.getCacheEntry().getGrantRecordsAsSecurable();
+    List<PolarisBaseEntity> granteeList = new ArrayList<>();
+    grantRecords.stream()
+        .map(
+            gr ->
+                entityCache.getOrLoadEntityById(
+                    callCtx, gr.getGranteeCatalogId(), gr.getGranteeId()))
+        .filter(lr -> lr != null && lr.getCacheEntry() != null)
+        .map(lr -> lr.getCacheEntry().getEntity())
+        .forEach(granteeList::add);
+    if (granteeList.size() != lookupResult.getCacheEntry().getGrantRecordsAsSecurable().size()) {
+      LOGGER.error(
+          "Failed to resolve all grantees for securable {}",
+          lookupResult.getCacheEntry().getEntity());
+      return new LoadGrantsResult(BaseResult.ReturnStatus.GRANT_NOT_FOUND, null);
+    }
+
+    // If the securable is the root container, then we need to add a grant record for the
+    // service_admin PrincipalRole, which is the only role that has the SERVICE_MANAGE_ACCESS
+    // privilege on the root
+    if (lookupResult
+        .getCacheEntry()
+        .getEntity()
+        .getName()
+        .equals(PolarisEntityConstants.getRootContainerName())) {
```

Comment:
Special cases by entity name look risky to me. The entity here is not restricted, right? So, it can be a namespace, I suppose, and might have the name of `root_container`, right?

Reply:
That's a good callout. This is intended to replicate the "simulated" root container entity logic that was in the `PolarisEntityManager`, but I'll add logic and a test to ensure we aren't adding privileges to random other entity types.